key.go

```diff
@@ -0,0 +1,777 @@
+
+	// Key-type-specific parameters. Which of these need to be set, and
+	// what their values ought to be, is determined by the KeyType.
+
+	// RawKeyParam contains the raw CBOR encoded data for the label -1.
+	// Depending on the KeyType this is used to populate either Curve or K
+	// below.
+	RawKeyParam cbor.RawMessage `cbor:"-1,keyasint,omitempty"`
+
+	// EC2/OKP params
+
+	// X is the x-coordinate
+	X []byte `cbor:"-2,keyasint,omitempty"`
+	// Y is the y-coordinate (sign bits are not supported)
+	Y []byte `cbor:"-3,keyasint,omitempty"`
+	// D is the private key
+	D []byte `cbor:"-4,keyasint,omitempty"`
+}
+
+// CreateOkpKey returns a Key created using the provided Octet Key Pair data.
+func CreateOkpKey(curve Curve, x, d []byte) (*Key, error) {
+	key := &Key{
+		Curve: curve,
+		keyStruct: keyStruct{
+			KeyType: KeyTypeOkp,
+			X:       x,
+			D:       d,
+		},
+	}
+	return key, key.Validate()
+}
+
+// CreateEc2Key returns a Key created using the provided elliptic curve key
+// data.
+func CreateEc2Key(curve Curve, x, y, d []byte) (*Key, error) {
+	key := &Key{
+		Curve: curve,
+		keyStruct: keyStruct{
+			KeyType: KeyTypeEc2,
+			X:       x,
+			Y:       y,
+			D:       d,
+		},
+	}
+	return key, key.Validate()
+
+}
+
+// CreateSymmetricKey returns a Key created using the provided Symmetric key
+// bytes.
+func CreateSymmetricKey(k []byte) (*Key, error) {
+	key := &Key{
+		K: k,
+		keyStruct: keyStruct{
+			KeyType: KeyTypeSymmetric,
+		},
+	}
+	return key, key.Validate()
+}
+
+// CreateKeyFromPublic returns a Key created using the provided crypto.PublicKey
+// and Algorithm.
+func CreateKeyFromPublic(alg Algorithm, pub crypto.PublicKey) (*Key, error) {
+	switch alg {
+	case AlgorithmES256, AlgorithmES384, AlgorithmES512:
+		vk, ok := pub.(*ecdsa.PublicKey)
+		if !ok {
+			return nil, fmt.Errorf("%v: %w", alg, ErrInvalidPubKey)
+		}
+
+		var curve Curve
+
+		switch alg {
+		case AlgorithmES256:
+			curve = CurveP256
+		case AlgorithmES384:
+			curve = CurveP384
+		case AlgorithmES512:
+			curve = CurveP521
+		}
+
+		return &Key{
+			Curve: curve,
+			keyStruct: keyStruct{
+				KeyType:   KeyTypeEc2,
+				Algorithm: alg,
+				X:         vk.X.Bytes(),
+				Y:         vk.Y.Bytes(),
+			},
+		}, nil
+	case AlgorithmEd25519:
+		vk, ok := pub.(ed25519.PublicKey)
+		if !ok {
+			return nil, fmt.Errorf("%v: %w", alg, ErrInvalidPubKey)
+		}
+
+		return &Key{
+			Curve: CurveEd25519,
+			keyStruct: keyStruct{
+				KeyType:   KeyTypeOkp,
+				Algorithm: alg,
+				X:         []byte(vk),
+			},
+		}, nil
+	default:
+		return nil, ErrAlgorithmNotSupported
+	}
+}
+
+// CreateKeyFromPrivate returns a Key created using provided crypto.PrivateKey
+// and Algorithm.
+func CreateKeyFromPrivate(alg Algorithm, pub crypto.PrivateKey) (*Key, error) {
+	switch alg {
+	case AlgorithmES256, AlgorithmES384, AlgorithmES512:
+		sk, ok := pub.(*ecdsa.PrivateKey)
+		if !ok {
+			return nil, fmt.Errorf("%v: %w", alg, ErrInvalidPrivKey)
+		}
+
+		var curve Curve
```

Comment:
computeCurveFromAlg() can be factored out

Reply:
not in a straight-forward way -- note that the cases here are restricted to those "caught" by the outer `switch`. Factoring this out means losing this context, and therefore needing to handle other curves as errors. 

IMO, it is better to have this simple `switch` repeated in two places, than deal with that.